tests/koji/test_koji_vmi.py

```diff
@@ -0,0 +1,226 @@
+import os
+
+from pytest import raises
+
+from pushsource import (
+    Source,
+    AmiPushItem,
+    VHDPushItem,
+    VMIPushItem,
+    AmiRelease,
+    VMIRelease,
+)
+
+
+def test_koji_vmi_notfound(fake_koji):
+    """koji source referencing nonexistent VMI build will fail."""
+
+    source = Source.get("koji:https://koji.example.com/?vmi_build=notexist-1.2.3")
+    with raises(ValueError) as exc_info:
+        list(source)
+
+    assert "Virtual machine image build not found in koji: notexist-1.2.3" in str(
+        exc_info.value
+    )
+
+
+def test_koji_not_vmi_invalid(fake_koji, koji_dir):
+    """Koji source referencing an invalid build will fail."""
+
+    source = Source.get(
+        "koji:https://koji.example.com/?vmi_build=foobuild-1.0-1", basedir=koji_dir
+    )
+
+    fake_koji.build_data[1234] = {
+        "id": 1234,
+        "name": "foobuild",
+        "version": "1.0",
+        "release": "1",
+        "nvr": "foobuild-1.0-1",
+        "completion_time": "2022-12-20 12:30:25",
+    }
+    fake_koji.build_data["foobuild-1.0-1"] = fake_koji.build_data[1234]
+
+    with raises(ValueError) as exc_info:
+        list(source)
+
+    assert (
+        "Build foobuild-1.0-1 not recognized as a virtual machine image build"
+        in str(exc_info.value)
+    )
+
+
+def test_koji_not_vmi_container(fake_koji, koji_dir):
+    """Koji source referencing a container will fail."""
+
+    source = Source.get(
+        "koji:https://koji.example.com/?vmi_build=foobuild-1.0-1", basedir=koji_dir
+    )
+
+    # Container image build
+    fake_koji.build_data[1234] = {
+        "id": 1234,
+        "name": "foobuild",
+        "version": "1.0",
+        "release": "1",
+        "nvr": "foobuild-1.0-1",
+        "completion_time": "2022-12-20 12:30:25",
+        "extra": {"typeinfo": {"image": {}}},
+        "container_koji_task_id": "container",
+    }
+    fake_koji.build_data["foobuild-1.0-1"] = fake_koji.build_data[1234]
+    with raises(ValueError) as exc_info:
+        list(source)
+
+    assert "The build foobuild-1.0-1 is for container images." in str(exc_info.value)
+
+
+def test_koji_vmis(fake_koji, koji_dir):
+    """Koji source yields requested Virtual Machine Images."""
+
+    nvr = "foobuild-1.0-1"
+    name = "foobuild"
+    version = "1.0"
+    release = "1"
```

Comment:
a nitpick
I would keep separate name version and release and merge them to nvr
```
nvr = "%s-%s-%s" % (name, version, release)
```
later can be changed to f-string after dropping support for old version of python
I would just like to avoid copy&paste issues/typos.

Reply:
Sure, I've fixed it now. Thanks!